### PR TITLE
Support globs for reload/restart actions

### PR DESCRIPTION
**Pull request by 9999years**
Instead of supplying paths to reload for and paths to restart for (and maybe some extra extensions to restart for), this lets you use standard glob expressions to indicate paths to restart, reload, or ignore.

This will close out several tickets.

### CODE REVIEW

**Comment by 9999years** [comment]
Current dependencies on/for this PR:
* main
  * **PR #118** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghciwatch/118" target="_blank">Graphite</a> 
    * **PR #97** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghciwatch/97" target="_blank">Graphite</a> 
      * **PR #138** <a href="https://app.graphite.dev/github/pr/MercuryTechnologies/ghciwatch/138" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/MercuryTechnologies/ghciwatch/138?utm_source=stack-comment).

File: src/ghci/mod.rs [hunk 1]
```diff
@@ -323,39 +318,56 @@ impl Ghci {
         for event in events {
             let path = event.as_path();
             let path = self.relative_path(path)?;
-            // Restart on `.cabal` files, `.ghci` files, and any paths under the `restart_paths`.
-            if path.extension().map(|ext| ext == "cabal").unwrap_or(false)
+
+            let restart_match = self.opts.restart_globs.matched(&path);
+            let reload_match = self.opts.reload_globs.matched(&path);
+            let path_is_haskell_source_file = is_haskell_source_file(&path);
```
Comment (parsonsmatt) [inline src/ghci/mod.rs:324]: I think we probably also want to modify the `is_haskell_source_file` impl to ignore hidden files. That'll fix the emacs bug by default.

File: tests/globs.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
+use test_harness::fs;
+use test_harness::test;
+use test_harness::BaseMatcher;
+use test_harness::GhciWatchBuilder;
+use test_harness::Matcher;
+
+/// Test that `ghcid-ng` can reload when a file matching a `--reload-glob` is changed.
```
Comment (parsonsmatt) [inline tests/globs.rs:7]:
```suggestion
/// Test that `ghciwatch` can reload when a file matching a `--reload-glob` is changed.
```

File: tests/globs.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        .await
+        .unwrap();
+
+    session
+        .wait_until_restart()
+        .await
+        .expect("ghcid-ng restarts when .cabal files change");
+}
+
+/// Test that `ghcid-ng` can restart when a `.ghci` file is changed.
```
Comment (parsonsmatt) [inline tests/globs.rs:101]:
```suggestion
/// Test that `ghciwatch` can restart when a `.ghci` file is changed.
```

File: tests/globs.rs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+    fs::touch(session.path(".ghci")).await.unwrap();
+
+    session
+        .wait_until_restart()
+        .await
+        .expect("ghcid-ng restarts when .ghci files change");
+}
+
+/// Test that `ghcid-ng` doesn't restart when a `.ghci` file is changed when `--restart-glob !.ghci`
```
Comment (parsonsmatt) [inline tests/globs.rs:124]:
```suggestion
/// Test that `ghciwatch` doesn't restart when a `.ghci` file is changed when `--restart-glob !.ghci`
```

File: tests/globs.rs [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+    fs::touch(session.path(".ghci")).await.unwrap();
+
+    session
+        .wait_for_log(BaseMatcher::reload_completes().but_not(BaseMatcher::restart()))
+        .await
+        .expect("ghcid-ng doesn't restart when ignored globs are changed");
+}
+
+/// Test that `ghcid-ng` restarts when a Haskell module is removed, even if a `--restart-glob`
```
Comment (parsonsmatt) [inline tests/globs.rs:148]:
```suggestion
/// Test that `ghciwatch` restarts when a Haskell module is removed, even if a `--restart-glob`
```

**Review by parsonsmatt — COMMENTED** [review]
I think we want to modify `is_haskell_source_file` otherwise all emacs users everywhere will need to specify a custom glob, which isn't great UX. Otherwise looks great!

**Review by parsonsmatt — APPROVED** [review]
awesome!!!!

**Pull request merged**